inferno-core/src/Inferno/Module/Prelude/Defs.hs

```diff
@@ -355,6 +355,16 @@ minFun = bimap (min) (bimap (min) (min))
 maxFun :: Either3 Int64 Double EpochTime -> Either3 (Int64 -> Int64) (Double -> Double) (EpochTime -> EpochTime)
 maxFun = bimap (max) (bimap (max) (max))
 
+arrayIndexFun :: (MonadThrow m, Pretty c) => Value c m
+arrayIndexFun =
+  VFun $ \case
+    VArray a -> pure $ VFun $ \v -> do
+      i <- fromValue v
+      if 0 <= i && i < length a
```

Comment:
Depending on the value of `i`, this could end up traversing the length of the list twice because `length` needs to traverse the whole list, and then `!!` also needs to traverse until at least `i`. I think it might be better to back the array type with a contiguous type, like a `Vector`, but that would be a much larger change of course, unrelated to this PR

Otherwise, you could use this [safe indexing function](https://hackage.haskell.org/package/extra-1.7.14/docs/src/Data.List.Extra.html#%21%3F) or write something equivalent to it to avoid traversing the list twice